Serialization Error When Trying to Write a COG Layer to S3

The following error is encountered when trying to save a COG layer to S3:

```
18/03/26 11:45:49 WARN AmazonS3Client: No content length specified for stream data.  Stream contents will be buffered in memory and could result in out of memory errors.
18/03/26 11:45:49 WARN AmazonS3Client: No content length specified for stream data.  Stream contents will be buffered in memory and could result in out of memory errors.
org.apache.spark.SparkException: Task not serializable
  at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:345)
  at org.apache.spark.util.ClosureCleaner$.org$apache$spark$util$ClosureCleaner$$clean(ClosureCleaner.scala:335)
  at org.apache.spark.util.ClosureCleaner$.clean(ClosureCleaner.scala:159)
  at org.apache.spark.SparkContext.clean(SparkContext.scala:2292)
  at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1.apply(RDD.scala:928)
  at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1.apply(RDD.scala:927)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
  at org.apache.spark.rdd.RDD.withScope(RDD.scala:363)
  at org.apache.spark.rdd.RDD.foreachPartition(RDD.scala:927)
  at geotrellis.spark.io.s3.cog.S3COGLayerWriter$$anonfun$writeCOGLayer$3.apply(S3COGLayerWriter.scala:85)
  at geotrellis.spark.io.s3.cog.S3COGLayerWriter$$anonfun$writeCOGLayer$3.apply(S3COGLayerWriter.scala:64)
  at scala.collection.TraversableLike$WithFilter$$anonfun$foreach$1.apply(TraversableLike.scala:733)
  at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
  at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
  at scala.collection.TraversableLike$WithFilter.foreach(TraversableLike.scala:732)
  at geotrellis.spark.io.s3.cog.S3COGLayerWriter.writeCOGLayer(S3COGLayerWriter.scala:64)
  at geotrellis.spark.io.cog.COGLayerWriter$class.write(COGLayerWriter.scala:78)
  at geotrellis.spark.io.s3.cog.S3COGLayerWriter.write(S3COGLayerWriter.scala:22)
  at geotrellis.spark.io.cog.COGLayerWriter$class.write(COGLayerWriter.scala:57)
  at geotrellis.spark.io.s3.cog.S3COGLayerWriter.write(S3COGLayerWriter.scala:22)
  ... 40 elided
Caused by: java.io.NotSerializableException: com.github.blemale.scaffeine.Cache
	- field (class "scala.Tuple2", name: "_2", type: "class java.lang.Object")
	- object (class "scala.Tuple2", (true,Cache(com.github.benmanes.caffeine.cache.BoundedLocalCache$BoundedLocalManualCache@4f4a4853)))
	- field (class "geotrellis.spark.io.s3.S3AttributeStore", name: "geotrellis$spark$io$AttributeCaching$$x$1", type: "class scala.Tuple2")
	- object (class "geotrellis.spark.io.s3.S3AttributeStore", geotrellis.spark.io.s3.S3AttributeStore@1e0f5681)
	- field (class "geotrellis.spark.io.s3.cog.S3COGLayerWriter", name: "attributeStore", type: "interface geotrellis.spark.io.AttributeStore")
	- object (class "geotrellis.spark.io.s3.cog.S3COGLayerWriter", geotrellis.spark.io.s3.cog.S3COGLayerWriter@7557ab7b)
	- field (class "geotrellis.spark.io.s3.cog.S3COGLayerWriter$$anonfun$writeCOGLayer$3", name: "$outer", type: "class geotrellis.spark.io.s3.cog.S3COGLayerWriter")
	- object (class "geotrellis.spark.io.s3.cog.S3COGLayerWriter$$anonfun$writeCOGLayer$3", <function1>)
	- field (class "geotrellis.spark.io.s3.cog.S3COGLayerWriter$$anonfun$writeCOGLayer$3$$anonfun$apply$3", name: "$outer", type: "class geotrellis.spark.io.s3.cog.S3COGLayerWriter$$anonfun$writeCOGLayer$3")
	- root object (class "geotrellis.spark.io.s3.cog.S3COGLayerWriter$$anonfun$writeCOGLayer$3$$anonfun$apply$3", <function1>)
  at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1182)
  at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
  at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
  at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
  at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
  at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
  at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
  at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
  at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
  at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
  at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
  at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
  at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
  at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
  at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
  at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
  at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
  at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
  at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
  at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
  at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
  at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
  at org.apache.spark.serializer.JavaSerializationStream.writeObject(JavaSerializer.scala:43)
  at org.apache.spark.serializer.JavaSerializerInstance.serialize(JavaSerializer.scala:100)
  at org.apache.spark.util.ClosureCleaner$.ensureSerializable(ClosureCleaner.scala:342)
  ... 60 more
```

This was the code that was being run:

```scala
object NLCDIngest extends CommandApp(

  name   = "nlcd-ingest",
  header = "Re-ingest the NLCD layer",
  main   = {

    /* Ensures that only positive, non-zero values can be given as arguments. */
    type UInt = Int Refined Positive

    val partO: Opts[UInt]    = Opts.option[UInt]("partitions", help = "Spark partitions to use.").withDefault(1600)

    (partO).map { (numPartitions) =>

      val conf = new SparkConf()
        .setIfMissing("spark.master", "local[*]")
        .setAppName("NLCD Ingest")
        .set("spark.serializer", classOf[KryoSerializer].getName)
        .set("spark.kryo.registrator", classOf[KryoRegistrator].getName)
        .set("spark.default.parallelism", "1600")

      implicit val ss = SparkSession.builder.config(conf).enableHiveSupport.getOrCreate
      implicit val sc = ss.sparkContext

      val store = S3AttributeStore("azavea-datahub", "catalog")
      val reader = S3LayerReader(store)
      val writer = S3COGLayerWriter(store)

      val layoutScheme: ZoomedLayoutScheme  = ZoomedLayoutScheme(WebMercator)

      try {
        val layer = reader.read[SpatialKey, Tile, TileLayerMetadata[SpatialKey]](LayerId("nlcd-2011-native", 0))

        val (z, reprojected) = layer.reproject(layoutScheme)

        val pyramidOptions = Pyramid.Options(partitioner = Some(ZPartitioner(1600)))

        val pyramid: Stream[(Int, TileLayerRDD[SpatialKey])] =
          Pyramid.levelStream(reprojected, layoutScheme, z, 0, pyramidOptions)

        val maxLayer = pyramid.maxBy(_._1)

        writer.write("nlcd-2011-epsg3857-cogs", maxLayer._2, maxLayer._1, ZCurveKeyIndexMethod)

      } finally {
        ss.stop
      }
    }
  }
)

--- a/.travis/hbase-install.sh
+++ b/.travis/hbase-install.sh
@@ -1,6 +1,6 @@
 #! /bin/bash
 
-if [ ! -f $HOME/downloads/hbase-1.3.1-bin.tar.gz ]; then sudo wget -O $HOME/downloads/hbase-1.3.1-bin.tar.gz http://www-eu.apache.org/dist/hbase/1.3.1/hbase-1.3.1-bin.tar.gz; fi
-sudo mv $HOME/downloads/hbase-1.3.1-bin.tar.gz hbase-1.3.1-bin.tar.gz && tar xzf hbase-1.3.1-bin.tar.gz
-sudo rm -f hbase-1.3.1/conf/hbase-site.xml && sudo mv .travis/hbase/hbase-site.xml hbase-1.3.1/conf
-sudo hbase-1.3.1/bin/start-hbase.sh
+if [ ! -f $HOME/downloads/hbase-1.3.2-bin.tar.gz ]; then sudo wget -O $HOME/downloads/hbase-1.3.2-bin.tar.gz http://www-eu.apache.org/dist/hbase/1.3.2/hbase-1.3.2-bin.tar.gz; fi
+sudo mv $HOME/downloads/hbase-1.3.2-bin.tar.gz hbase-1.3.2-bin.tar.gz && tar xzf hbase-1.3.2-bin.tar.gz
+sudo rm -f hbase-1.3.2/conf/hbase-site.xml && sudo mv .travis/hbase/hbase-site.xml hbase-1.3.2/conf
+sudo hbase-1.3.2/bin/start-hbase.sh

--- a/spark/src/main/scala/geotrellis/spark/io/AttributeCaching.scala
+++ b/spark/src/main/scala/geotrellis/spark/io/AttributeCaching.scala
@@ -25,18 +25,14 @@ import com.typesafe.config.ConfigFactory
 import scala.concurrent.duration._
 
 trait AttributeCaching { self: AttributeStore =>
-  @transient private lazy val (enabled, cache) = {
-    val config = ConfigFactory.load()
-    val expiration = config.getInt("geotrellis.attribute.caching.expirationMinutes")
-    val maxSize = config.getInt("geotrellis.attribute.caching.maxSize")
-    val enabled = config.getBoolean("geotrellis.attribute.caching.enabled")
+  import AttributeCaching._
 
-    enabled -> Scaffeine()
+  @transient private lazy val cache =
+    Scaffeine()
       .recordStats()
       .expireAfterWrite(expiration.minutes)
       .maximumSize(maxSize)
       .build[(LayerId, String), Any]
-  }
 
   def cacheRead[T: JsonFormat](layerId: LayerId, attributeName: String): T = {
     if(enabled)
@@ -64,4 +60,11 @@ trait AttributeCaching { self: AttributeStore =>
   def clearCache(id: LayerId, attribute: String): Unit = {
     if(enabled) cache.invalidate(id -> attribute)
   }
+}
+
+object AttributeCaching extends Serializable {
+  lazy val config     = ConfigFactory.load()
+  lazy val expiration = config.getInt("geotrellis.attribute.caching.expirationMinutes")
+  lazy val maxSize    = config.getInt("geotrellis.attribute.caching.maxSize")
+  lazy val enabled    = config.getBoolean("geotrellis.attribute.caching.enabled")
 }

